File: wabi/update-zcode.py
#!/usr/bin/env python3
import urllib.request
import re
import sys
import subprocess
import os

DESKTOP_NIX = os.path.join(os.path.dirname(os.path.dirname(__file__)), "modules/hosts/apostrophe/packages/desktop.nix")

def get_latest_zcode_version():
    url = "https://zcode.z.ai/en"
    try:
        req = urllib.request.Request(
            url, 
            headers={'User-Agent': 'Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36'}
        )
        with urllib.request.urlopen(req, timeout=5) as response:
            html = response.read().decode('utf-8')
    except Exception as e:
        print(f"Error fetching latest version from {url}: {e}")
        return None

    # Find matches like: https://cdn-zcode.z.ai/zcode/electron/releases/3.2.5/ZCode-3.2.5-linux-x64.AppImage
    pattern = r'https://cdn-zcode\.z\.ai/zcode/electron/releases/(\d+\.\d+\.\d+)/ZCode-\1-linux-x64\.AppImage'
    versions = re.findall(pattern, html)
    if not versions:
        print("Could not find any ZCode release versions in page.")
        return None

    # Sort versions semantically
    def semver_key(v):
        return [int(x) for x in v.split('.')]

    latest_version = max(set(versions), key=semver_key)
    return latest_version

def get_current_zcode_info():
    if not os.path.exists(DESKTOP_NIX):
        print(f"Error: {DESKTOP_NIX} not found.")
        return None, None

    with open(DESKTOP_NIX, 'r') as f:
        content = f.read()

    # Search for ZCode block
    zcode_match = re.search(r'# -- ZCode --.*?pname\s*=\s*"zcode";\s*version\s*=\s*"([^"]+)";.*?sha256\s*=\s*"([^"]+)";', content, re.DOTALL)
    if not zcode_match:
        print("Could not locate ZCode block in desktop.nix.")
        return None, None

    return zcode_match.group(1), zcode_match.group(2)

def update_zcode_in_file(new_version, new_hash):
    with open(DESKTOP_NIX, 'r') as f:
        content = f.read()

    # Regex to find the version, url, and sha256 within the ZCode block
    pattern = re.compile(
        r'(# -- ZCode --.*?pname\s*=\s*"zcode";\s*version\s*=\s*")[^"]+("\s*;\s*src\s*=\s*fetchurl\s*\{\s*url\s*=\s*")[^"]+("\s*;\s*sha256\s*=\s*")[^"]+("\s*;\s*\};)',
        re.DOTALL
    )

    new_url = f"https://cdn-zcode.z.ai/zcode/electron/releases/{new_version}/ZCode-{new_version}-linux-x64.AppImage"

    def repl(m):
        return f"{m.group(1)}{new_version}{m.group(2)}{new_url}{m.group(3)}{new_hash}{m.group(4)}"

    new_content, count = pattern.subn(repl, content)
    if count == 0:
        print("Failed to replace ZCode block in desktop.nix.")
        return False

    with open(DESKTOP_NIX, 'w') as f:
        f.write(new_content)

    print(f"Successfully updated desktop.nix to version {new_version} with hash {new_hash}")
    return True

def main():
    print("Checking for ZCode updates...")
    latest_version = get_latest_zcode_version()
    if not latest_version:
        sys.exit(1)

    current_version, current_hash = get_current_zcode_info()
    if not current_version:
        sys.exit(1)

    print(f"Current local version: {current_version}")
    print(f"Latest online version: {latest_version}")

    if latest_version == current_version:
        print("ZCode is already up to date!")
        sys.exit(0)

    print(f"New version {latest_version} available! Fetching new hash...")
    new_url = f"https://cdn-zcode.z.ai/zcode/electron/releases/{latest_version}/ZCode-{latest_version}-linux-x64.AppImage"
    
    try:
        # Run nix-prefetch-url to get the new sha256 hash in base32
        result = subprocess.run(["nix-prefetch-url", new_url], capture_output=True, text=True, check=True)
        new_hash = result.stdout.strip()
        if not new_hash:
            print("Failed to get hash from nix-prefetch-url.")
            sys.exit(1)
        print(f"Fetched hash: {new_hash}")
    except Exception as e:
        print(f"Error running nix-prefetch-url: {e}")
        sys.exit(1)

    if update_zcode_in_file(latest_version, new_hash):
        print("Update complete!")
    else:
        sys.exit(1)

if __name__ == "__main__":
    main()

File: wabi/src/bin/utils/update-zcode.rs
use std::fs;
use std::path::Path;
use std::process::Command;

fn extract_field(content: &str, start_pattern: &str, end_pattern: &str) -> Option<String> {
    let start_idx = content.find(start_pattern)?;
    let val_start = start_idx + start_pattern.len();
    let end_idx = content[val_start..].find(end_pattern)?;
    Some(content[val_start..val_start + end_idx].to_string())
}

fn parse_version(v: &str) -> Vec<u32> {
    v.split('.').filter_map(|s| s.parse::<u32>().ok()).collect()
}

fn get_latest_zcode_version() -> Option<String> {
    let output = Command::new("curl")
        .args([
            "-s",
            "-H",
            "User-Agent: Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36",
            "https://zcode.z.ai/en",
        ])
        .output()
        .ok()?;

    if !output.status.success() {
        eprintln!("Failed to fetch ZCode website using curl.");
        return None;
    }

    let html = String::from_utf8_lossy(&output.stdout);
    let pattern = "https://cdn-zcode.z.ai/zcode/electron/releases/";
    let mut versions = Vec::new();
    let mut search_str = &*html;

    while let Some(idx) = search_str.find(pattern) {
        let start = idx + pattern.len();
        let remaining = &search_str[start..];
        if let Some(slash_idx) = remaining.find('/') {
            let version = &remaining[..slash_idx];
            if version.chars().all(|c| c.is_ascii_digit() || c == '.') && !version.is_empty() {
                versions.push(version.to_string());
            }
        }
        search_str = remaining;
    }

    if versions.is_empty() {
        eprintln!("Could not find any ZCode release versions on page.");
        return None;
    }

    // Sort versions semantically and return the highest
    versions.sort_by_key(|v| parse_version(v));
    versions.last().cloned()
}

fn main() -> Result<(), Box<dyn std::error::Error>> {
    println!("Checking for ZCode updates...");

    let desktop_nix_path = Path::new("modules/hosts/apostrophe/packages/desktop.nix");
    if !desktop_nix_path.exists() {
        eprintln!("Error: {:?} not found.", desktop_nix_path);
        std::process::exit(1);
    }

    let content = fs::read_to_string(desktop_nix_path)?;

    // Locate ZCode block
    let start_idx = content
        .find("# -- ZCode --")
        .ok_or("Cannot find ZCode header")?;
    let zcode_block = &content[start_idx..];

    let current_version =
        extract_field(zcode_block, "version = \"", "\";").ok_or("Cannot find current version")?;
    let current_hash =
        extract_field(zcode_block, "sha256 = \"", "\";").ok_or("Cannot find current hash")?;

    println!("Current local version: {}", current_version);

    let latest_version = match get_latest_zcode_version() {
        Some(v) => v,
        None => std::process::exit(1),
    };

    println!("Latest online version: {}", latest_version);

    if latest_version == current_version {
        println!("ZCode is already up to date!");
        return Ok(());
    }

    println!(
        "New version {} available! Fetching new hash...",
        latest_version
    );
    let new_url = format!(
        "https://cdn-zcode.z.ai/zcode/electron/releases/{}/ZCode-{}-linux-x64.AppImage",
        latest_version, latest_version
    );

    let output = Command::new("nix-prefetch-url").arg(&new_url).output()?;

    if !output.status.success() {
        eprintln!("Failed to get hash from nix-prefetch-url.");
        std::process::exit(1);
    }

    let new_hash = String::from_utf8_lossy(&output.stdout).trim().to_string();
    if new_hash.is_empty() {
        eprintln!("nix-prefetch-url returned an empty hash.");
        std::process::exit(1);
    }
    println!("Fetched hash: {}", new_hash);

    // Re-read file to avoid race conditions and replace the block
    let content = fs::read_to_string(desktop_nix_path)?;
    let start_idx = content
        .find("# -- ZCode --")
        .ok_or("Cannot find ZCode header")?;
    let end_idx = content[start_idx..]
        .find("})")
        .ok_or("Cannot find end of ZCode block")?
        + start_idx
        + 2;
    let old_block = &content[start_idx..end_idx];

    let mut new_block = old_block.to_string();
    new_block = new_block.replace(
        &format!("version = \"{}\";", current_version),
        &format!("version = \"{}\";", latest_version),
    );
    new_block = new_block.replace(
        &format!("releases/{}/", current_version),
        &format!("releases/{}/", latest_version),
    );
    new_block = new_block.replace(
        &format!("ZCode-{}-linux", current_version),
        &format!("ZCode-{}-linux", latest_version),
    );
    new_block = new_block.replace(
        &format!("sha256 = \"{}\";", current_hash),
        &format!("sha256 = \"{}\";", new_hash),
    );

    let mut new_content = content.clone();
    new_content.replace_range(start_idx..end_idx, &new_block);
    fs::write(desktop_nix_path, new_content)?;

    println!(
        "Successfully updated desktop.nix to version {} with hash {}",
        latest_version, new_hash
    );
    println!("Update complete!");

    Ok(())
}
